Hashlink.js

```diff
@@ -34,8 +35,65 @@ export class Cipher {
    *
    * @returns {Promise<string>} Resolves to a string that is a hashlink.
    */
-  async create({data, urls, transforms, meta}) {
-    throw new Error('Not implemented.');
+  async create({data, urls, transforms, meta = {}}) {
+    // ensure data or urls are provided
+    if(data === undefined && urls == undefined) {
+      throw new Error('Either `data` or `urls` must be provided.')
+    }
+
+    // transforms are provided
+    if(transforms === undefined) {
+      throw new Error('The hashlink creation `transforms` must be provided.')
+    }
+
+    if(urls) {
+      // ensure urls are an array
+      if(Array.isArray(urls) === false) {
+        urls = [urls];
+      }
+
+      // ensure all URLs are strings
+      urls.forEach(url => {
+        if(typeof url !== 'string') {
+          throw new Error('Each `url` must be a string. Value: ' + url, url);
+        }
+      });
+    }
+
+    // merge meta options with urls
+    meta = Object.assign(meta, {'url': urls});
+
+    // generate the encoded cryptographic hash
+    const outputData = await transforms.reduce(
+      async (transformedData, transform) => {
+      transformedData = await transformedData;
+      transformedData = await this.registeredTransforms[transform].encode(
+        transformedData);
+      return transformedData;
+    }, data);
+
+    // generate the encoded metadata
+    const metadata = new Map();
+    if(meta.url) {
+      metadata.set(0x0f, meta.url);
+    }
+    if(meta['content-type']) {
+      metadata.set(0x0e, meta['content-type']);
+    }
+    if(meta.experimental) {
+      metadata.set(0x0d, meta.experimental);
+    }
+    let mhMeta = '';
+    if(metadata.size > 0) {
+      const baseEncodingTransform = transforms[transforms.length - 1];
+      const cborData = new Uint8Array(cbor.encode(metadata));
+      const mbCborData = new TextDecoder().decode(
+        this.registeredTransforms[baseEncodingTransform].encode(cborData));
+
+      mhMeta = ':' + mbCborData;
+    }
+
+    return 'hl:' + new TextDecoder().decode(outputData) + mhMeta;
```

Comment:
Could reformulate this as:

```js
    const textDecoder = new TextDecoder();
    let hashlink = 'hl:' + textDecoder.decode(outputData);

    // append meta data if present
    if(metadata.size > 0) {
      const baseEncodingTransform = transforms[transforms.length - 1];
      const cborData = new Uint8Array(cbor.encode(metadata));
      const mbCborData = textDecoder.decode(
        this.registeredTransforms[baseEncodingTransform].encode(cborData));
      hashlink += ':' + mbCborData;
    }

    return hashlink;
```

Reply:
Simplified in ebd661c1fcaceff5eccd5269caa818ff134c9af5